### PR TITLE
Migrate from boost::optional to std::optional in EventSetup system

### PR DESCRIPTION
**Pull request by knoepfel**
This PR *does* include some whitespace changes which are due to using C++17's nested namespace declarations.

### CODE REVIEW

**Comment by andrius-k** [comment]
please test

File: DQMServices/Components/src/DQMDaqInfo.cc [hunk 1]
```diff
@@ -11,7 +11,7 @@ void DQMDaqInfo::beginLuminosityBlock(const edm::LuminosityBlock& lumiBlock, con
   
   edm::eventsetup::EventSetupRecordKey recordKey(edm::eventsetup::EventSetupRecordKey::TypeTag::findType("RunInfoRcd"));
   
-  if( nullptr != iSetup.find( recordKey ) ) {
+  if( std::nullopt != iSetup.find( recordKey ) ) {
```
Comment (Dr15Jones) [inline DQMServices/Components/src/DQMDaqInfo.cc]: Should this also be changed to use `tryToGet`?

File: FWCore/Framework/src/ESProducer.cc [hunk 2]
```diff
@@ -2,7 +2,7 @@
 //
 // Package:     Framework
 // Class  :     ESProducer
-// 
+//
```
Comment (Dr15Jones) [inline FWCore/Framework/src/ESProducer.cc]: It looks like you only changed spaces here? Can this be reverted back?

File: FWCore/Framework/interface/produce_helpers.h [hunk 3]
```diff
@@ -23,88 +23,85 @@
...
-namespace edm {
-   namespace eventsetup {
-      
-      namespace produce { struct Null;}
-      
-     template<typename FromT, typename ToT> void moveFromTo(FromT& iFrom,
-                                                            ToT & iTo) {
-       iTo = std::move(iFrom);
-     }
+namespace edm::eventsetup {
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/produce_helpers.h]: Was this just a change from 3 spaces to 2 ? If so, please revert.

File: FWCore/Framework/interface/es_Label.h [hunk 4]
```diff
@@ -29,71 +30,75 @@
...
+  };
+
+  template<int ILabel, typename T>
+  L<T,ILabel> l(std::shared_ptr<T>& iP) {
+    return L<T,ILabel>{iP};
+  }
+
+  struct Label {
+    Label() = default;
+    Label(std::string_view const iString) : default_{iString} {}
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/es_Label.h]: Was `string_view` the only change?

File: FWCore/Framework/test/DummyProxyProvider.h [hunk 5]
```diff
@@ -4,7 +4,7 @@
 //
 // Package:     Framework
 // Class  :     DummyProvider
-// 
+//
```
Comment (Dr15Jones) [inline FWCore/Framework/test/DummyProxyProvider.h]: Were only cosmetic changes done to this file?

File: FWCore/Framework/test/DummyData.h [hunk 6]
```diff
@@ -11,15 +11,11 @@
...
-namespace edm {
-   namespace eventsetup {
-      namespace test {
-         struct DummyData { int value_;
-            DummyData(int iValue=0) : value_(iValue) {}
-            void dummy() {} // Just to suppress compilation warning message
-         };
-      }
-   }
+namespace edm::eventsetup::test {
```
Comment (Dr15Jones) [inline FWCore/Framework/test/DummyData.h]: Is this only cosmetic changes?

File: FWCore/Framework/interface/ESProducts.h [hunk 7]
```diff
@@ -28,114 +28,113 @@
 
 // forward declarations
 namespace edm {
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/ESProducts.h]: Are the only changes the spacing and the collapsed namespace? If so, please revert.

File: FWCore/Framework/interface/ESProducer.h [hunk 8]
```diff
@@ -4,7 +4,7 @@
 //
 // Package:     Framework
 // Class  :     ESProducer
-//
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/ESProducer.h]: Were there any functional changes to this file?

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

File: DQMServices/Components/src/DQMDaqInfo.cc (hunk 1 above)
Comment (knoepfel) [reply DQMServices/Components/src/DQMDaqInfo.cc]: Yes...I missed that one.  Will change.

File: FWCore/Framework/interface/ESProducer.h (hunk 8 above)
Comment (knoepfel) [reply FWCore/Framework/interface/ESProducer.h]: I think the answer is no.  I'll be more careful next time.

File: FWCore/Framework/interface/ESProducts.h (hunk 7 above)
Comment (knoepfel) [reply FWCore/Framework/interface/ESProducts.h]: Yes, will revert.

File: FWCore/Framework/interface/es_Label.h (hunk 4 above)
Comment (knoepfel) [reply FWCore/Framework/interface/es_Label.h]: That and the namespace collapsing.

File: FWCore/Framework/interface/produce_helpers.h (hunk 3 above)
Comment (knoepfel) [reply FWCore/Framework/interface/produce_helpers.h]: Namespace collapsing was the change.

File: FWCore/Framework/src/ESProducer.cc (hunk 2 above)
Comment (knoepfel) [reply FWCore/Framework/src/ESProducer.cc]: Yikes, yes that's true.  Will revert.

File: FWCore/Framework/test/DummyData.h (hunk 6 above)
Comment (knoepfel) [reply FWCore/Framework/test/DummyData.h]: Yes

File: FWCore/Framework/test/DummyProxyProvider.h (hunk 5 above)
Comment (knoepfel) [reply FWCore/Framework/test/DummyProxyProvider.h]: Yes

File: L1TriggerConfig/RCTConfigProducers/src/L1RCTChannelMaskTester.cc [hunk 9]
```diff
@@ -63,7 +63,7 @@ void L1RCTChannelMaskTester::analyze(const edm::Event& iEvent,
             edm::eventsetup::EventSetupRecordKey::TypeTag::findType(
                     "L1RCTNoisyChannelMaskRcd"));
 
-    if (evSetup.find(recordKey) == nullptr) {
+    if (evSetup.find(recordKey) == std::nullopt) {
         //record not found
```
Comment (knoepfel) [inline L1TriggerConfig/RCTConfigProducers/src/L1RCTChannelMaskTester.cc]: Missed this one, too--should be `tryToGet`.

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Framework/interface/es_Label.h (hunk 4 above)
Comment (Dr15Jones) [reply FWCore/Framework/interface/es_Label.h]: I think the switch to std::string_view is causing the compilation problem in HcalHardcodeGeometryEP.

Comment (knoepfel) [reply FWCore/Framework/interface/es_Label.h]: Yes, just confirmed that.  Will revert.

**Comment by andrius-k** [comment]
please test

**Comment by andrius-k** [comment]
+1

**Comment by Dr15Jones** [comment]
+1
The histogram failures are all from L1TEMU/L1TStage2CaloLayer2/L1TdeStage2CaloLayer2 which I believe has a history of false positives.

**Comment by Dr15Jones** [comment]
@monttj @arizzi @gpetruc @ggovi @rekovic @thomreis @nsmith- 
please review and sign

**Comment by thomreis** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**